### PR TITLE
Document the selectors for the wc/store/checkout data store

### LINKED ISSUE #6945: Write external documentation for the `wc/store/checkout` data store
The data stores in Gutenberg have their selectors and actions documented on pages. Example: https://developer.wordpress.org/block-editor/reference-guides/data/data-core-editor/

We should ensure the actions and selectors of the new data stores are documented too. This documentation should go in `docs/third-party-developers/data-stores`.

We need to ensure we document:
- Each of the available selectors
- What they return and when, i.e. how each selector's return value will change depending on the state of the checkout.
- The actions and the arguments supplied to each action and how each action will affect the state of the checkout.
- We could also provide some examples of how to do "common" tasks, i.e. show an example of a component that renders an `input` field. Use the data store to find out if the checkout is processing, and if so sets the `input` to `readonly`. There are many more examples that we could do here too, feel free to combine multiple actions or selectors into one larger example too.

Issue #6771 will be used to track the _internal_ documentation for this data store.

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #6945

## Notes

In #6612, we ran a Data Store Migration. This PR aims to create the external documentation for the `wc/store/checkout` data store.

### Testing

1. Open https://github.com/woocommerce/woocommerce-blocks/blob/361b48b6220f57fab8a62b9a9272757530d6f14a/assets/js/data/checkout/selectors.ts
2. Verify that `checkout.md` from this PR documents all selector that appear in the document mentioned before.

* [x] Do not include in the Testing Notes

### CODE REVIEW

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/checkout.md]: Perhaps "Whether the checkout is processing" is slightly better,  WC orders have a "processing" status don't they and this could potentially confuse people

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/checkout.md]: I think we shouldn't document this, we will be removing it

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/checkout.md]: Is this correct? I think this is true when there's an in-flight request to update these values, isn't it?

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/checkout.md]: I think that this is true while the checkout isn't processing or calculating, right? Cc @alexflorisca to verify

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (nielslange) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: Just removed it from the docs.

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (nielslange) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: I'm not 100% sure about that. That function is defined here: https://github.com/woocommerce/woocommerce-blocks/blob/361b48b6220f57fab8a62b9a9272757530d6f14a/assets/js/data/checkout/selectors.ts#L43-L45

And `calculatingCount` holds the following comment: https://github.com/woocommerce/woocommerce-blocks/blob/361b48b6220f57fab8a62b9a9272757530d6f14a/assets/js/data/checkout/default-state.ts#L15-L16

That's why I thought that this function tells whether some calculations in the checkout are required. That said, I might have worded it poorly. I'll change it according to your suggestion.

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (nielslange) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: Thanks for your suggestion. I've updated the wording accordingly.

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (nielslange) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: I'll keep this open to wait for @alexflorisca feedback.

Comment (alexflorisca) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: I think a more accurate description of this might be "Checkout has had some activity, but is currently waiting for user input". Basically what happens is when you first load checkout block, the status is `pristine`. When any action is called on this store (so basically when. the user does anything in the UX), we change the status to idle. Once the user clicks "Place Order", we change the status to "before_processing", "processing" and "after_processing" and maybe back to "idle" if there were errors.

Comment (nielslange) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: Thanks for the suggestion, @alexflorisca. I just changed the wording accordingly.

**Review by alexflorisca — COMMENTED** [review]
Nice work @nielslange, this looks good but let's wait to make a decision on how we continue with the data store work based on pdFofs-fT-p2

**Comment by nielslange** [comment]
> Nice work @nielslange, this looks good but let's wait to make a decision on how we continue with the data store work based on https://woocheckout.wordpress.com/2022/09/06/data-architecture-for-cart-checkout-block/

Sure, @alexflorisca.

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/checkout.md]: Could we specify that this is the WordPress user ID of the customer whose order the Checkout block is currently processing?

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/checkout.md]: Could we change this to be something like "Whether the checkout is processing an existing order" cc @alexflorisca for wording ideas here

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (opr) [inline docs/third-party-developers/extensibility/data-store/checkout.md]:
```suggestion
Returns true if the order is complete, and false otherwise.

#### _Returns_

-   `boolean`: Whether the order is complete.
```

**Review by opr — CHANGES_REQUESTED** [review]
Thanks for documenting these selectors, Niels! Just a couple of comments and we should be good to go 👏🏼

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (nielslange) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: Thanks for pointing this out, @opr. I just adjusted the wording according to your suggestion.

File: docs/third-party-developers/extensibility/data-store/checkout.md
Comment (alexflorisca) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: It's tricky because in most of the descriptions here you are saying what the return value is. So I think we should change the descriptions to be more generic (for example here, it could be "Has a draft order been created?" and the return value description to something like "true if order exists, false otherwise". You can look [here](https://github.com/woocommerce/woocommerce-blocks/blob/feature/data-store-refactor/assets/js/data/checkout/default-state.ts#L12-L34) for inspiration with the descriptions

Comment (nielslange) [reply docs/third-party-developers/extensibility/data-store/checkout.md]: Thanks for your reply, @alexflorisca. In https://github.com/woocommerce/woocommerce-blocks/issues/6945, @opr
mentioned https://developer.wordpress.org/block-editor/reference-guides/data/data-core-editor/ as an example. For consistency-reasons, I followed their style. This is the way they document `has***` functions:

> **hasChangedContent**
>
> Returns true if content includes unsaved changes, or false otherwise.
> 
> _Parameters_
> 
> - state **Object**: Editor state.
> 
> _Returns_
> 
> - **boolean**: Whether content includes unsaved changes.

With that in mind, I changed the description from `Returns true if an order exists, and false otherwise.` to `Returns true if a draft order had been created, and false otherwise.` and the return value description from `boolean: Whether an order occurred.` to `boolean: Whether a draft order had been created.`.

Would you mind doing another review?

**Review by alexflorisca — APPROVED** [review]
Yeah, thanks for making those changes Niels, makes sense to leave it as you have it. Good work, everything looks good 👍 ~~Let's wait to merge this until we merge the data store work into trunk though otherwise it will document things that aren't yet there!~~  Would you mind changing the branch that it gets merged into from `trunk` to `feature/data-store-refactor`

**Comment by nielslange** [comment]
> Yeah, thanks for making those changes Niels, makes sense to leave it as you have it. Good work, everything looks good 👍 ~Let's wait to merge this until we merge the data store work into trunk though otherwise it will document things that aren't yet there!~ Would you mind changing the branch that it gets merged into from `trunk` to `feature/data-store-refactor`

Thanks for your review, @alexflorisca. I just rebased this PR from `trunk` to `feature/data-store-refactor` and merged it.

**Pull request merged**
